### DISCUSSION TITLE
Better treeshaking of ajax methods: Is it ok to import from `rxjs/internal/*`?

### DISCUSSION THREAD

**Discussion by rart** [ReactiveX/rxjs#5994] [Q&A] [answered]
To use rxjs' ajax methods, it looks like you're supposed to `import { ajax } from 'rxjs/ajax'`, right?
Is it ok to `import { ajaxGet } from 'rxjs/internal/observable/dom/AjaxObservable'` instead?
If you're producing a bundle (e.g. with rollup) and you're only using `ajaxGet` for example, it'd be nice not to have the whole rest of the ajax package in. Why aren't the ajax methods published by `rxjs/ajax`?

In general, is `internal` considered private stuff that you're not supposed to import directly? Or is it ok to import there from?

**Comment by kwonoj** [comment, marked as answer]
`internal` as name stands for it is not public entrypoint. Importing non-public subpath is not gauranteed by semvor and can break anytime.

>  it'd be nice not to have the whole rest of the ajax package in.

what is imported when you treeshake with importing ajax? ajax helper method / others are thin wrapper to ajaxObservable so in most cases it'll be needed anyway.

**Reply by rart** [reply]
Thanks for your reply.

> what is imported when you treeshake with importing ajax? 

ajaxGet, ajaxPost, ajaxDelete, ajaxPut, ajaxPatch are all included despite only using ajaxGet. I guess this is due to the `AjaxObservable.create` method? It makes it all unshakeable.

It does look like the only shakeable thing would be those small wrapper methods (ajaxGet, ajaxPost, etc), which I guess isn't much.

Since AjaxObservable is a private, would it makes sense to strip out the `AjaxObservable.create` out of AjaxObservable for this small gain in _modularity_?

**Reply by kwonoj** [reply]
> ajaxGet, ajaxPost, ajaxDelete, ajaxPut, ajaxPatch are all included 

Yes, but as said it's just near-1 line wrapper:

https://github.com/ReactiveX/rxjs/blob/032f6815779bd7e3c0f2317748043b084bee01d3/src/internal/ajax/ajax.ts#L150-L152

I'm curious how large amount of byte we are dealing with.

In theory we could export all as individual modules but we take some balances instead of exposing all. I don't think we'll try to split down everything at this moment.
